libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1132 @@
+}
+
+static gboolean
+flatpak_ref_equal (gconstpointer a, gconstpointer b)
+{
+  return g_str_equal (flatpak_ref_get_name (FLATPAK_REF (a)),
+                      flatpak_ref_get_name (FLATPAK_REF (b)));
+}
+
+/* Squash actions on the same ref into the last action on that ref, returning
+ * a pointer array of remote ref actions, ordered by the order key in each
+ * remote ref action. */
+static GPtrArray *
+squash_ref_actions_ptr_array (GPtrArray *ref_actions)
+{
+  g_autoptr(GHashTable) hash_table = g_hash_table_new_full (flatpak_ref_hash,
+                                                            flatpak_ref_equal,
+                                                            g_object_unref,
+                                                            (GDestroyNotify) flatpak_remote_ref_action_unref);
+  GPtrArray *squashed_ref_actions = NULL;
+  gsize i;
+  gpointer key;
+  gpointer value;
+  GHashTableIter hash_iter;
+
+  for (i = 0; i < ref_actions->len; ++i)
+    {
+      FlatpakRemoteRefAction *action = g_ptr_array_index (ref_actions, i);
+      g_hash_table_replace (hash_table,
+                            g_object_ref (action->ref),
+                            flatpak_remote_ref_action_ref (action));
+    }
+
+  squashed_ref_actions = g_ptr_array_new_full (g_hash_table_size (hash_table),
+                                               (GDestroyNotify) flatpak_remote_ref_action_unref);
+  g_hash_table_iter_init (&hash_iter, hash_table);
+
+  while (g_hash_table_iter_next (&hash_iter, &key, &value))
+    g_ptr_array_add (squashed_ref_actions,
+                     flatpak_remote_ref_action_ref (value));
+
+  g_ptr_array_sort (squashed_ref_actions, sort_flatpak_remote_ref_actions);
+
+  return squashed_ref_actions;
+}
+
+/* Given a hash table of filenames to FlatpakRemoteRefActionsFile, hoist
+ * the underlying pointer array of remote ref actions and make that the value
+ * of the new hash table.
+ *
+ * This will make the hash table suitable for passing to
+ * eos_updater_util_squash_remote_ref_actions */
+GHashTable *
+eos_updater_util_hoist_flatpak_remote_ref_actions (GHashTable *ref_actions_file_table)
+{
+  g_autoptr(GHashTable) hoisted_ref_actions_table = g_hash_table_new_full (g_str_hash,
+                                                                           g_str_equal,
+                                                                           g_free,
+                                                                           (GDestroyNotify) g_ptr_array_unref);
+  gpointer key, value;
+  GHashTableIter iter;
+
+  g_hash_table_iter_init (&iter, ref_actions_file_table);
+
+  while (g_hash_table_iter_next (&iter, &key, &value))
+    {
+      FlatpakRemoteRefActionsFile *ref_actions_file = value;
+
+      g_hash_table_insert (hoisted_ref_actions_table,
+                           g_strdup (key),
+                           g_steal_pointer (&ref_actions_file->remote_ref_actions));
```

Comment:
I don't like live-modifying the hash table pointers like this; you should be able to simply use `g_ptr_array_ref()` instead of stealing the pointer.